extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -936,7 +936,12 @@ private List<SegmentIdWithShardSpec> generateSegmentIdsWithShardSpecsForAppend(
         }
       }
 
-      if (allocation == null) {
+      // Even if allocation isn't null, the overlord makes the best effort job of allocating a segment with the given
+      // segmentGranularity. This is commonly seen in case when there is already a coarser segment in the interval where
+      // the requested segment is present and that segment completely overlaps the request. Throw an error if the interval
+      // doesn't match the granularity requested
+      if (allocation == null
+          || !IntervalUtils.doesIntervalMatchesGranularity(allocation.getInterval(), segmentGranularity)) {
```

Comment:
I think the error message should be different in the null case and the unexpected interval case.

Reply:
Noted! I think we can make a new fault class then and signify to the user the pre-existing segment so that the user can use `REPLACE`, change the granularity in the query or not insert in the given time interval